api/services/syncing/syncing.go

```diff
@@ -260,78 +302,267 @@ func (ss *StateSync) GetConsensusHashes() bool {
 	}
 	wg.Wait()
-	Log.Info("syncing: Finished downloadBlocks.")
+	Log.Info("[sync] Finished downloadBlocks.")
+}
+
+// CompareBlockByHash compares two block by hash, it will be used in sort the blocks
+func CompareBlockByHash(a *types.Block, b *types.Block) int {
+	ha := a.Hash()
+	hb := b.Hash()
+	return bytes.Compare(ha[:], hb[:])
+}
+
+// GetHowManyMaxConsensus will get the most common blocks and the first such blockID
+func GetHowManyMaxConsensus(blocks []*types.Block) (int, int) {
+	// As all peers are sorted by their blockHashes, all equal blockHashes should come together and consecutively.
+	curCount := 0
+	curFirstID := -1
+	maxCount := 0
+	maxFirstID := -1
+	for i := range blocks {
+		if curFirstID == -1 || CompareBlockByHash(blocks[curFirstID], blocks[i]) != 0 {
+			curCount = 1
+			curFirstID = i
+		} else {
+			curCount++
+		}
+		if curCount > maxCount {
+			maxCount = curCount
+			maxFirstID = curFirstID
+		}
+	}
+	return maxFirstID, maxCount
+}
+
+func (ss *StateSync) getMaxConsensusBlockFromParentHash(parentHash common.Hash) *types.Block {
+	candidateBlocks := []*types.Block{}
+	ss.syncMux.Lock()
+	for id := range ss.syncConfig.peers {
+		peerConfig := ss.syncConfig.peers[id]
+		for _, block := range peerConfig.newBlocks {
+			ph := block.ParentHash()
+			if bytes.Compare(ph[:], parentHash[:]) == 0 {
+				candidateBlocks = append(candidateBlocks, block)
+				break
+			}
+		}
+	}
+	ss.syncMux.Unlock()
+	if len(candidateBlocks) == 0 {
+		return nil
+	}
+	// Sort by blockHashes.
+	sort.Slice(candidateBlocks, func(i, j int) bool {
+		return CompareBlockByHash(candidateBlocks[i], candidateBlocks[j]) == -1
+	})
+	maxFirstID, maxCount := GetHowManyMaxConsensus(candidateBlocks)
+	Log.Debug("[sync] Find block with matching parenthash", "parentHash", parentHash, "hash", candidateBlocks[maxFirstID].Hash(), "maxCount", maxCount)
+	return candidateBlocks[maxFirstID]
+}
+
+func (ss *StateSync) getBlockFromOldBlocksByParentHash(parentHash common.Hash) *types.Block {
+	for _, block := range ss.commonBlocks {
+		ph := block.ParentHash()
+		if bytes.Compare(ph[:], parentHash[:]) == 0 {
+			return block
+		}
+	}
+	return nil
+}
+
+func (ss *StateSync) getBlockFromLastMileBlocksByParentHash(parentHash common.Hash) *types.Block {
+	for _, block := range ss.lastMileBlocks {
+		ph := block.ParentHash()
+		if bytes.Compare(ph[:], parentHash[:]) == 0 {
+			return block
+		}
+	}
+	return nil
+}
+
+func (ss *StateSync) updateBlockAndStatus(block *types.Block, bc *core.BlockChain, worker *worker.Worker) bool {
+	Log.Info("[sync] Current Block", "blockHex", bc.CurrentBlock().Hash().Hex())
+	_, err := bc.InsertChain([]*types.Block{block})
+	if err != nil {
+		Log.Debug("Error adding new block to blockchain", "Error", err)
+		return false
+	}
+	Log.Info("[sync] new block [POST]added to blockchain", "blockHeight", bc.CurrentBlock().NumberU64(), "blockHex", bc.CurrentBlock().Hash().Hex(), "parentHex", bc.CurrentBlock().ParentHash().Hex())
```

Comment:
nit: missing a space after [POST], the statement is a bit unclear.

Reply:
it is used for distinguish [pre]; but I removed [POST] for clarity